Arbitrary ConditionSet Expressions

tehhowch opened the pull request:
This is one possible implementation of arbitrary condition support for conversation `branch` and missions' `to offer` and `to fail` (and assignment in conversation `apply` and mission `on offer`, `on complete`, etc)
With this PR
 - A condition can be more than 3 nodes long.
 - `%`, `*`, `*=`, `/`, and `/=` are valid operators (modulus has equal priority as multiply and divide).
 - parentheses can be used
    - They must be standalone: `2 * ( 1 + "rep: Republic" )` and not `2 * (1 + "rep: Republic")`
 - A ConditionSet can define variables (temporary if in a `branch`, `to offer`, or `to fail`) to hold results of computations, and will no longer erroneously fail `Test()` when using a mix of comparison and assignment expressions.

An example `to offer`
```
to offer
	"random" < ( 5 + "ships: Heavy Warship" ) / 2
	"completed U orders" = "Mining, Small [6]: done" + ... - ... * ... / ...
	100 * "reputation: Republic" <= "reputation: Syndicate" / ( "reputation: Free Worlds" + "reputation: Empire" )
	"random" < "completed U orders" + "random"
	year % 7 > 4
```
This PR achieves #2799 and #883, and enables such things as #3076, #2514

endless-sky commented:
It seems to me that a much better approach would be to work on supporting more complex condition expressions. I'm worried that "temporary" changes that don't permanently affect the conditions could be really confusing to content creators. And, creating a copy of the entire condition set (which could be quite large) every time a "to offer" is checked (which could be quite often) could end up being a performance issue as more missions get added.

tehhowch commented:
I hadn't considered the cost of duplicating the condition map - the save I use for testing has a condition map that has 1473 elements and I'm sure anyone who uses mods witl have double or even triple this.

My goal was to achieve this without really altering the meat of the conditionset test and apply functions, but it is going to be much less copy intensive to tell the conditionset if it has a mix of statement types during load, and then using a different test method if true (one that first searches a small map of the results from apply, then searches the passed const map if the string value wasnt found)

Regarding confusion - is the test method called anywhere except via mission branch checks and when checking if the mission can offer or can fail? I would argue that the behavior with this PR in those situations is less confusing than the existing behavior, in which no error messages are thrown if assignment operators are used, and the test may fail unexpectedly.

As far as starting on more complex expression, handling, i feel the use of temporary variables is one part of that goal.

tehhowch commented:
Profiled the landing behavior for my test save by taking off & landing 20 times
<details><summary>8a3dfea </summary>

 
[8a3dfea.txt](https://github.com/endless-sky/endless-sky/files/1389082/8a3dfea.txt)
![8a3dfea](https://user-images.githubusercontent.com/20871346/31635463-6caf516a-b28c-11e7-8097-9114078a9c59.png)
</details>

<details><summary>048ac19e260b18179101f6a36ae285a7fb8a3c7b (master) </summary>


[master.txt](https://github.com/endless-sky/endless-sky/files/1389092/master.txt)
![master](https://user-images.githubusercontent.com/20871346/31635623-071c7bb0-b28d-11e7-816e-1050b29e6dd1.png)
</details>

<details><summary>5aefd5d </summary>


[5aefd5d.txt](https://github.com/endless-sky/endless-sky/files/1389102/5aefd5d.txt)
![5aefd5d](https://user-images.githubusercontent.com/20871346/31635796-be857fcc-b28d-11e7-8d8e-1b0a923b1588.png)
</details>

This can probably be cleaned up a bit more, since I don't think there will be any significant issues if I drop the overloads for TokenValue and TestSet and let every `Test()` call reference the (maybe empty) map first.

tehhowch commented:
I've extended this PR to support arbitrary-length expressions by decomposing expressions from a datanode into incremental expressions if the standard `Add` methods have failed.

Another way to do it would be to alter the structure of the Expression class.

tehhowch commented:
Marking as WIP since this PR, as-is, will save the temporary conditions to the player's save if a complex condition is used in an `apply` or `on _____ ` node. Probably best to not set a bunch of `_t_{digits}` conditions.

tehhowch commented:
Added support for arbitrary conditions without generating a bunch of temporary variables. Also added support for parentheses.

Still no support for exponentiation, but because temporary conditions can be defined, it can be done easily for integer exponents:
```
to offer
    "var" = ( "thing 1" + "thing 2" - "thing 3" * "thing 4" ... ) + ... * ... / ...
    "var" * "var" <= "something"
```
if exponentiating in an `apply` node (e.g. `on offer`, `on complete`, `apply`), an additional operation can be added to remove the temporary variable after it has served its purpose.
```
on offer
    "var" = ( "thing 1" + "thing 2" - "thing 3" * "thing 4" ... ) + ... * ... / ...
    "new condition" >?= "var" * "var" * "var"
    clear var
```

Amazinite commented:
Would you be able to handle #3217 in this PR?

tehhowch commented:
@Amazinite this PR should now support the `%` operator. Feel free to build it and verify your conditions

tehhowch commented:
@endless-sky Do you want me to work on incorporating #2456 features as well? Or would you prefer that to be a follow-up PR?
If text substitutions should be in a follow-up, then this PR is ready for your review.

tehhowch commented:
To help content creators use this PR's features, I've updated the changes from #2530 to use the supported syntax and thus provide an example usage. This does slightly change the offer rates from those posted in #2530, to the following:

P(offer) \ assists | 0 | 1 | 2 | 3
-- | -- | -- | -- | -- 
recon | 35% | 40% | 55% | 80%
katya 1 | 10% | 17% | 39% | 75%
katya alt 2 |   | 25% | 50% | 75%
escort |   | 20% | 40% | 60%
bounty 1 | 20% | 25% | 40% | 65%
*bounty 2 | 25% | 30% | 45% | 70%
*bounty 3 | 30% | 35% | 50% | 75%
total | 56% | 61% | 70% | 71%

Other than the FW intro missions, there currently aren't any jobs/missions that are offered more commonly if you've done more of them (i.e. due to being trusted more by that faction), but future storylines and /or faction developments (e.g. #3120 ) probably will use this mechanic.

jafdy commented:
I don't like the exact way the numbers come out there (and something feels wrong with the sums) but it's not too bad. I'll throw a spreadsheet at it again at some point if this gets merged.

VonZorch commented:
I don't use GitHub much.  How do I use this with the current master file?

tehhowch commented:
Not sure what you mean. Given that there are no conflicts that would need to be resolved between this branch and es-master, adding my fork as a remote and then using the appropriate `git pull` command should suffice. The details of those procedures should be adequately covered by online git tutorials. If not, please use Steam / Google Groups / Discord for more responsive help. I recommend deleting your post (after which I will delete this one), as it is not really related to this PR's features.

After testing the features added by this PR, if you have feedback to share, please share it here :+1:

petervdmeer commented:
Didn't check the details, but this looks like a very good improvement.
I don't expect difficult conflicts between this change and feature request #4308.

petervdmeer commented on source/ConditionSet.cpp:
```diff
@@ -40,15 +42,183 @@ namespace {
+	}
+	
+	// Ensure the ConditionSet line has balanced parentheses on both sides.
+	bool HasUnbalancedParentheses(const vector<string> &tokens)
```
Is it possible to fully parse the expression (each and every token, including variables and sub-expressions) using a parsing technique like for example https://en.wikipedia.org/wiki/Recursive_descent_parser#C_implementation?

Your implementation supports quite some different expressions, but if such full parsing is possible, then I expect that the game can support a lot more complex expressions while at the same time also providing a stronger verification of the correctness of the expressions.

tehhowch replied:
I'm still quite green when it comes to algorithms so if a more complex parser implementation is preferred I'll have to ask you propose the changes. You can either open a PR to this branch (on my fork) or a similar PR to `master` here.

petervdmeer replied:
I did some work on parsers a very long time ago, so I'm willing to give it a try.

What I usually do for complex changes is to split them into multiple smaller PRs and only start working on the next PR when the previous one is tested, reviewed and merged. For example:
* PR #4391 prepares the code, data-files and savegames for additional bay-types, but the actual change to enable different bay-types will be a separate PR.
* PR #4370 prepares the framework for auto-testing, but I have a list of about 7 different PRs that I should create to get to really good auto-tests.
Would this way of working also work for Endless Sky? Or does the merge-process require a different way of working?

tehhowch commented on source/ConditionSet.cpp:
```diff
@@ -40,15 +42,183 @@ namespace {
+		return false;
+	}
+	
+	// Converts the given vector of strings into a vector of 64-bit ints.
```
This function handles the run-time (as opposed to load-time) conversion of the strings into the values they will have when passed to the associated operators. This is where e.g. `"reputation: Republic"` is converted into e.g. `1034` (or whatever your current reputation with the Republic government happens to be). Rather than interleave reading the condition `map`s with performing the actual condition math, the maps are used, and then the 2 vectors are used.

Any suggestions of a better comment?

petervdmeer replied:
Something like: Looks up the values of the variables that are stored as strings in the first vector and stores those values in the second vector as 64-bit int values.

tehhowch commented on source/ConditionSet.cpp:
```diff
@@ -105,41 +271,78 @@ bool ConditionSet::IsEmpty() const
+				}
+				else if(!op.empty())
+					rhs.emplace_back(token);
+				else if(IsComparison(token))
```
The bracket checks are done before reaching this stage, in the check for condition validity

petervdmeer replied:
Ok

tehhowch commented on source/ConditionSet.cpp:
```diff
@@ -40,15 +42,183 @@ namespace {
+	{
+		static const set<string> invalids = {
+			"{", "}", "[", "]", "|", "^", "&", "!", "~",
+			"||", "&&", "&=", "|=", "<<", ">>"
```
I suppose I could change things around and just check that expressions are always separated by valid operators. I don't expect there are other seemingly valid mathematical operators that many would try to use so I just listed them here.

petervdmeer replied:
I'm affraid that content creators will surprise us with the combinations that they come up with.

petervdmeer commented:
For when we pick this up: We could check if we can generate the parser using a tool like https://en.wikipedia.org/wiki/Yacc instead of building the complete parsing ourselves.

Pointedstick commented:
Needs a rebase BTW

tehhowch commented:
Yeah I'm not worrying about this until we've shipped 9.10 :)

LocalGod79 commented:
i'd like to chime in (after bouncing to this from a steam thread) - M.Z.'s right about trying to keep things simple for developers - the feature is nice, but wiki documentation needs to follow if you're adding this.
